text/0005-is-operator.md

```diff
@@ -0,0 +1,176 @@
+
+4. The `is` operator may encourage coding practices that go against our recommended "best practices" (more details below).
+
+## Best practice: Using "typed" actions
+
+From the discussion on <https://github.com/cedar-policy/cedar/issues/94>:
+
+> Guidance we have given other customers is that actions should be "typed," in the sense that they include in their name the type of the resource they apply to. So instead of writing
+>
+>     permit(
+>       principal == User::"admin",
+>       action == Action::"view",
+>       resource is Photo
+>     );
+>
+> you would write
+>
+>     permit(
+>       principal == User::"admin",
+>       action == Action::"viewPhoto",
+>       resource
+>     );
+>
+> In your schema, you indicate that the "viewPhoto" action should always be accompanied by a `Photo` for the resource, so the policy doesn't need to say that explicitly (and doesn't need to perform a runtime check).
+
+If we support an `is` operator, then customers may be encouraged to use a single "view" action for multiple resource types, rather than a new "viewX" actions for each resource type X.
+
+# Alternatives
+
+1. For the most part, users can currently solve the problems that `is` solves by specifying in the schema what types of principals and resources an action applies to, or by using `has` checks, per the examples in the [Motivation](#motivation) above.
+However, in some cases, `is` may result in policies that are easier to read.
+(Aside: the issue of "input validation" brought up in [Use case 1](#use-case-1-input-validation) could also be addressed by performing validation against the schema for every input query, but that deserves its own RFC.)
+
+2. Another way to simulate `is` is to add an `entity_type` attribute to entities, and check the type of an entity using `resource.entity_type == "File"` (for example).
+However, this requires users to manually add an entity's type as an attribute, and leads to the possibility that an entity might be created with an attribute that doesn't actually match its type.
+
+3. The behavior of `is` can also be simulated using groups.
+For each entity type (e.g., `File`), a user could create a group (e.g., `Type::"File"`) and add all entities of that type to the group.
+Then to check whether an entity has a particular type, users could use `in` (e.g., `resource in Type::"File"`).
+Like alternative (2), this requires some extra work on the part of the user (setting up the memberOf relation), and does not prevent entities from being added as members of the wrong group.
+
+# Unresolved questions
+
+1. Should we allow `is` to apply to action entities? Action entities should _always_ have type `Action`, although they may have different namespaces. So `is` could still be used to distinguish between `Namespace1::Action` and `Namespace2::Action`.
```

Comment:
I wouldn't. IMO `action is` probably shouldn't be written, but it isn't actively harmful. It would also be a pain to distinguish between `Action` entities and `AnyEntities` which could be actions, but `is` really should apply to `AnyEntity`.

There's precedent for different allowed conditions in the policy scope, so, if others want to restrict `is` on action I would be in favor of restricting `action` constraint in the scope only.

Reply:
Agreeing with @john-h-kastner-aws. We already have special rules in the scope, but in the condition clauses actions are first class entities and should remain as such.

Reply:
If we don't allow `is` in the policy scope at all (same rule as `has`), then @john-h-kastner-aws 's suggestion is vacuously satisfied.

Reply:
I'll take the other opinion, that we should allow `is` to apply to action entities.  Outside of the schema, Cedar doesn't require that action entities always have type `Action`, and I think `action is` might be useful for schema-less users.  (And perhaps, in the future, we might relax the restriction for `Action` types even for schema users, in a back-compatible way; I'm not suggesting that here, but we could at least leave the door open.)

Reply:
Re-reading, I see that actually I agree with @john-h-kastner-aws and @aaronjeline, and everyone is proposing allowing `action is` (at least outside the policy scope).

Reply:
In the latest revision posted, `action is` is allowed anywhere outside the scope.